registry/storage/driver/s3-aws/s3.go

```diff
@@ -665,103 +665,104 @@ func (d *driver) Reader(ctx context.Context, path string, offset int64) (io.Read
-			ContentType:          d.getContentType(),
-			ACL:                  d.getACL(),
-			ServerSideEncryption: d.getEncryptionMode(),
-			SSEKMSKeyId:          d.getSSEKMSKeyID(),
-			StorageClass:         d.getStorageClass(),
-		})
+		uploadID, err := d.createMultipartUpload(ctx, key)
 		if err != nil {
 			return nil, err
 		}
-		return d.newWriter(ctx, key, *resp.UploadId, nil), nil
+		return d.newWriter(ctx, key, *uploadID, nil), nil
 	}
 
-	listMultipartUploadsInput := &s3.ListMultipartUploadsInput{
-		Bucket: aws.String(d.Bucket),
-		Prefix: aws.String(key),
+	uploadID, err := d.inProgressUpload(ctx, path)
+	if err != nil {
+		return nil, err
 	}
-	for {
-		resp, err := d.S3.ListMultipartUploadsWithContext(ctx, listMultipartUploadsInput)
+
+	if uploadID == nil {
+		uploadID, err := d.createMultipartUpload(ctx, key)
 		if err != nil {
-			return nil, parseError(path, err)
+			return nil, err
 		}
+		return d.newWriter(ctx, key, *uploadID, nil), nil
+	}
 
-		// resp.Uploads can only be empty on the first call
-		// if there were no more results to return after the first call, resp.IsTruncated would have been false
-		// and the loop would be exited without recalling ListMultipartUploads
-		if len(resp.Uploads) == 0 {
-			fi, err := d.Stat(ctx, path)
-			if err != nil {
-				return nil, parseError(path, err)
-			}
+	parts, err := d.listParts(ctx, path, uploadID)
+	if err != nil {
+		return nil, err
+	}
+	return d.newWriter(ctx, key, *uploadID, parts), nil
+}
 
-			if fi.Size() == 0 {
-				resp, err := d.S3.CreateMultipartUploadWithContext(ctx, &s3.CreateMultipartUploadInput{
-					Bucket:               aws.String(d.Bucket),
-					Key:                  aws.String(key),
-					ContentType:          d.getContentType(),
-					ACL:                  d.getACL(),
-					ServerSideEncryption: d.getEncryptionMode(),
-					SSEKMSKeyId:          d.getSSEKMSKeyID(),
-					StorageClass:         d.getStorageClass(),
-				})
-				if err != nil {
-					return nil, err
-				}
-				return d.newWriter(ctx, key, *resp.UploadId, nil), nil
-			}
-			return nil, storagedriver.Error{
-				DriverName: driverName,
-				Detail:     fmt.Errorf("append to zero-size path %s unsupported", path),
-			}
-		}
+func (d *driver) createMultipartUpload(ctx context.Context, key string) (*string, error) {
+	res, err := d.S3.CreateMultipartUploadWithContext(ctx, &s3.CreateMultipartUploadInput{
+		Bucket:               aws.String(d.Bucket),
+		Key:                  aws.String(key),
+		ContentType:          d.getContentType(),
+		ACL:                  d.getACL(),
+		ServerSideEncryption: d.getEncryptionMode(),
+		SSEKMSKeyId:          d.getSSEKMSKeyID(),
+		StorageClass:         d.getStorageClass(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return res.UploadId, nil
+}
 
-		var allParts []*s3.Part
-		for _, multi := range resp.Uploads {
-			if key != *multi.Key {
-				continue
-			}
+func (d *driver) inProgressUpload(ctx context.Context, path string) (uploadID *string, err error) {
```

Comment:
Please don't use named returns unless it significantly improves the code (it often doesn't and makes readers of the code miss important logic). Besides, you don't even take advantage of them through empty `return`s: all the returns in this func use explicit var names.

Reply:
Named returns also serve as a form of documentation. The meaning of `*string` without the named return is unclear.


```go
func (d *driver) inProgressUpload(ctx context.Context, path string) (uploadID *string, err error)
```

Reply:
> The meaning of *string without the named return is unclear.

Sure, but this is an unexported func, so one needs to read it anyway.

As I said, they _generally_ make it "harder" to read the code, which is why we've been removing them from this codebase.

Reply:
Do you feel in this case the named return parameter makes the code harder to understand? I'm not sure I see the benefit of:

```go
func (d *driver) inProgressUpload(ctx context.Context, path string) (*string, error) {
	var (
		uploadID *string
		empty    bool
	)
```

compared to

```go
func (d *driver) inProgressUpload(ctx context.Context, path string) (uploadID *string, err error) {
```

Reply:
I think we could compare our opinions on code documentation using named returns all day -- @milosgajdos is stating that precedent has been set in the repo to not use named returns so it makes sense to just follow suite here.